add safe_eq

rlkelly opened the pull request:
this is adds a method "_safe_eq" that returns a `Result` when comparing two ast nodes and checks that they have the same `SortKind`.

sameer commented on z3/src/ast.rs:
```diff
@@ -196,6 +196,25 @@ pub trait Ast<'ctx>: Sized + fmt::Debug + Into<Dynamic<'ctx>> {
+        if (left_sort == right_sort) {
+            Ok(self._eq(other))
+        } else {
+            Err(format!("Can not compare nodes, Sort does not match.  Nodes contain types {:?} and {:?}", left_sort, right_sort))
```
I agree that it's implied that the two don't have the same sort but I don't think the Err type should be a string. Maybe better suited as a struct whose display trait impl is the error message. Like

```rust
struct SortDiffers<'ctx> {
  left: Sort<'ctx>, right: Sort<'ctx>
}

impl<'ctx> Display for Sort<'ctx> { ... }
```

sameer commented on z3/src/ast.rs:
```diff
@@ -196,6 +196,25 @@ pub trait Ast<'ctx>: Sized + fmt::Debug + Into<Dynamic<'ctx>> {
+
+    /// Compare this `Ast` with another `Ast`, and get a Result.  Errors if the sort does not
+    /// match for the two values.
+    fn _safe_eq(&self, other: &Self) -> Result<Bool<'ctx>, std::string::String> {
```
now that there's a safe method could you make `_eq` call it? That way z3 can unwrap with the error message instead of just the opaque assertion.

rlkelly commented on z3/src/lib.rs:
```diff
@@ -17,7 +17,7 @@ extern crate num;
 use std::sync::Mutex;
 use z3_sys::*;
-pub use z3_sys::GoalPrec;
+pub use z3_sys::{AstKind, GoalPrec, SortKind};
```
I just want to point this out, I hope this is alright.

rlkelly commented:
#129

rlkelly commented on z3/src/ast.rs:
```diff
@@ -319,16 +337,23 @@ pub trait Ast<'ctx>: Sized + fmt::Debug + Into<Dynamic<'ctx>> {
+                let guard = Z3_MUTEX.lock().unwrap();
+                let app = Z3_to_app(ctx.z3_ctx, self.get_z3_ast());
+                let func_decl = Z3_get_app_decl(ctx.z3_ctx, app);
+                assert!(!func_decl.is_null(), "Ast has a null Z3_func_decl.");
```
it was a sanity check, we can get rid of it though

rlkelly commented:
I added three small changes.  I removed clone from tactic since this will raise reference count errors, made tactics take references, and made `list_subgoals` move the `ApplyResult`.

sameer commented:
Will review once I'm done with work but the changes sound reasonable 🙂

On Wed, Mar 10, 2021, at 17:10, Rob Kelly wrote:
> 

> this is ready for merge if the changes I made are acceptable


> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/prove-rs/z3.rs/pull/130#issuecomment-796207335>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACUVICENZHYFUKW7PW6Q5CTTC7N3RANCNFSM4YV2VLDA>.

sameer commented on z3/src/tactic.rs:
```diff
@@ -4,43 +4,25 @@ use std::result::Result;
-}
-
-impl<'ctx> Clone for Tactic<'ctx> {
-    fn clone(&self) -> Self {
```
I understand the point of the change and I'm fine with it, but in the future it would be better-suited to a separate PR

sameer reviewed (approved):
thanks for powering through these changes, looks great! :rocket:
